### PR TITLE
Support id column mapping for Delta Lake

### LINKED ISSUE #13629: Add support for `delta.columnMapping.mode='id'` in Delta Lake connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #13629

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Delta Lake
* Support reading tables with the property `delta.columnMapping.mode=id`. ({issue}`13629`)
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 1]
```diff
@@ -65,26 +65,16 @@ public void testColumnMappingModeNone()
     @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS})
     public void testColumnMappingModeId()
     {
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:67]: https://docs.databricks.com/delta/delta-column-mapping.html

>  It also allows users to name Delta table columns by using characters that are not allowed by Parquet, such as spaces, so that users can directly ingest CSV or JSON data into Delta without the need to rename columns due to previous character constraints.

Let's add a test case showcasing this functionality.

https://docs.databricks.com/delta/delta-column-mapping.html#supported-characters-in-column-names

> When column mapping is enabled for a Delta table, you can include spaces as well as any of these characters in the table’s column names: ,;{}()\n\t=.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 2]
```diff
@@ -325,9 +331,21 @@ && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parq
...
+
+    private static Optional<org.apache.parquet.schema.Type> getParquetType(GroupType groupType, ParquetColumnMapping columnMapping, HiveColumnHandle column)
+    {
+        if (columnMapping == FIELD_ID) {
+            return Optional.ofNullable(getParquetTypeById(column.getBaseHiveColumnIndex(), groupType));
+        }
+        if (columnMapping == NAME) {
             return Optional.ofNullable(getParquetTypeByName(column.getBaseColumnName(), groupType));
         }
         if (column.getBaseHiveColumnIndex() < groupType.getFieldCount()) {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:351]: Is this the case for `ParquetColumnMapping.COLUMN_INDEX` ?

Would it make sense to use a `switch` statement here ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 3]
```diff
@@ -390,8 +391,8 @@ public DeltaLakeTableHandle getTableHandle(ConnectorSession session, SchemaTable
         Optional<MetadataEntry> metadata = metastore.getMetadata(tableSnapshot, session);
         if (metadata.isPresent()) {
             ColumnMappingMode columnMappingMode = getColumnMappingMode(metadata.get());
-            if (columnMappingMode != ColumnMappingMode.NAME && columnMappingMode != ColumnMappingMode.NONE) {
-                throw new TrinoException(NOT_SUPPORTED, format("Only 'name' or 'none' is supported for the '%s' table property", COLUMN_MAPPING_MODE_CONFIGURATION_KEY));
+            if (columnMappingMode != ColumnMappingMode.ID && columnMappingMode != ColumnMappingMode.NAME && columnMappingMode != ColumnMappingMode.NONE) {
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:394]: Writing to such table is unsupported for now.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 4]
```diff
@@ -348,24 +366,35 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
-        }
-        else {
-            physicalName = fieldName;
-            physicalColumnType = columnType;
+        switch (mappingMode) {
+            case ID:
+                String columnMappingId = node.get("metadata").get("delta.columnMapping.id").asText();
+                verify(!isNullOrEmpty(columnMappingId), "id is null or empty");
+                fieldId = OptionalInt.of(Integer.parseInt(columnMappingId));
+                // Databricks stores column statics with physical name
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:377]: It was my typo. It's "column statistics".

**Comment by ebyhr** [comment]
@findinpath Addressed comments.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 5]
```diff
@@ -325,21 +331,35 @@ && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parq
...
+                return Optional.ofNullable(getParquetTypeByName(column.getBaseColumnName(), groupType));
+            case COLUMN_INDEX:
+                if (column.getBaseHiveColumnIndex() < groupType.getFieldCount()) {
+                    return Optional.of(groupType.getType(column.getBaseHiveColumnIndex()));
+                }
 
-        return Optional.empty();
+                return Optional.empty();
+        }
+        throw new IllegalArgumentException("Unsupported column mapping: " + columnMapping);
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:357]: nit: for consistency sake, use the `default` case in switch (as it is done in the rest of the codebase).

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 6]
```diff
@@ -309,10 +315,10 @@ && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parq
                 internalFields.add(Optional.empty());
             }
             else {
-                internalFields.add(Optional.ofNullable(getParquetType(column, fileSchema, useColumnNames))
+                internalFields.add(Optional.ofNullable(getParquetType(column, fileSchema, columnMapping))
                         .flatMap(field -> {
-                            String columnName = useColumnNames ? column.getBaseColumnName() : fileSchema.getFields().get(column.getBaseHiveColumnIndex()).getName();
-                            return constructField(column.getBaseType(), lookupColumnByName(messageColumn, columnName));
+                            ColumnIO columnIo = getColumnIo(column, messageColumn, fileSchema, columnMapping);
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:320]: preexisting: I find `field` variable name rather misleading because the method actually returns a type.

The variable `field` is actually fully ignored. 
Does it make sense to refactor the code to make it more understandable?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 7]
```diff
@@ -348,24 +366,35 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
         Type physicalColumnType;
-        if (mappingMode == ColumnMappingMode.NAME) {
-            physicalName = node.get("metadata").get("delta.columnMapping.physicalName").asText();
-            verify(!isNullOrEmpty(physicalName), "physicalName is null or empty");
-            physicalColumnType = buildType(typeManager, typeNode, true);
-        }
-        else {
-            physicalName = fieldName;
-            physicalColumnType = columnType;
+        switch (mappingMode) {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:372]: nit(optional): IDE suggests a Java 17 enhanced switch

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 8]
```diff
@@ -65,26 +65,16 @@ public void testColumnMappingModeNone()
...
-        }
-        finally {
-            onDelta().executeQuery("DROP TABLE default." + tableName);
-        }
+        testColumnMappingMode("id");
     }
 
     @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS})
     public void testColumnMappingModeName()
+    {
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:73]: Can we please have a rename test for a partitioning column ?

I want to make sure that the reading from the delta table still works after doing the rename with the `delta.column.mapping` set to `id` / `name`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java [hunk 9]
```diff
@@ -57,6 +58,7 @@
 
     private final String name;
     private final Type type;
+    private final OptionalInt fieldId;
     // Hold field names in Parquet files
     // The value is same as 'name' when the column mapping mode is none
     // The value is same as 'delta.columnMapping.physicalName' when the column mapping mode is name. e.g. col-6707cc9e-f3aa-4e6b-b8ef-1b03d3475680
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java:64]: Do we need comment updates for the fields `physicaName` and `phyisicalType`  ?

```
The value is same as 'delta.columnMapping.physicalName' when the column mapping mode is name. e.g. col-6707cc9e-f3aa-4e6b-b8ef-1b03d3475680
```
```
The value is same as 'delta.columnMapping.physicalName' when the column mapping mode is name. e.g. row(col-5924c8b3-04cf-4146-abb5-2c229e7ff708 integer)
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java [hunk 10]
```diff
@@ -151,7 +161,7 @@ public HiveColumnHandle toHiveColumnHandle()
     {
         return new HiveColumnHandle(
                 physicalName, // this name is used for accessing Parquet files, so it should be physical name
-                0, // hiveColumnIndex; we provide fake value because we always find columns by name
+                fieldId.orElse(0), // hiveColumnIndex; we provide fake value when delta.columnMapping.mode table property isn't id
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java:164]: I am confused here: Is `fieldId` the same as  hive column index?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 11]
```diff
@@ -321,8 +339,8 @@ public static List<ColumnMetadata> extractColumnMetadata(MetadataEntry metadataE
     public static List<DeltaLakeColumnMetadata> extractSchema(MetadataEntry metadataEntry, TypeManager typeManager)
     {
         ColumnMappingMode mappingMode = getColumnMappingMode(metadataEntry);
-        if (mappingMode != ColumnMappingMode.NAME && mappingMode != ColumnMappingMode.NONE) {
-            throw new TrinoException(NOT_SUPPORTED, format("Only 'name' or 'none' is supported for the '%s' table property", COLUMN_MAPPING_MODE_CONFIGURATION_KEY));
+        if (mappingMode != ColumnMappingMode.ID && mappingMode != ColumnMappingMode.NAME && mappingMode != ColumnMappingMode.NONE) {
```
Comment (homar) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:342]: It is the second place where exactly the same check is made. Maybe it would make sense to extract it ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 12]
```diff
@@ -325,21 +331,35 @@ && predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parq
...
+    {
+        switch (columnMapping) {
+            case FIELD_ID:
+                return Optional.ofNullable(getParquetTypeById(column.getBaseHiveColumnIndex(), groupType));
+            case NAME:
+                return Optional.ofNullable(getParquetTypeByName(column.getBaseColumnName(), groupType));
+            case COLUMN_INDEX:
+                if (column.getBaseHiveColumnIndex() < groupType.getFieldCount()) {
+                    return Optional.of(groupType.getType(column.getBaseHiveColumnIndex()));
+                }
```
Comment (homar) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:353]:
```
CASE UNKNOW:
CASE DEFAULT:
   return Optional.empty();
```
Maybe it'd be better to stay consistent ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 5 above)
Comment (homar) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:357]: +1

**Review by homar — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 5 above)
Comment (ebyhr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:357]: We shouldn't use `default` in this case. See https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-default-clause-in-exhaustive-enum-based-switch-statements and #6883

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:372]: I avoided it intentionally because there're other `switch` that can use enhanced version in this class and I didn't want to mix up in class.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java:164]: No, but it's used to provide the position in `HiveColumnHandle`.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 8 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:73]: Added a partition column and rename statement.

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java [hunk 13]
```diff
@@ -185,6 +185,17 @@ public static org.apache.parquet.schema.Type getParquetTypeByName(String columnN
...
 
+    public static org.apache.parquet.schema.Type getParquetTypeById(int fieldId, GroupType groupType)
+    {
+        for (org.apache.parquet.schema.Type type : groupType.getFields()) {
+            if (type.getId().intValue() == fieldId) {
+                return type;
+            }
+        }
+
+        return null;
```
Comment (alexjo2144) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:196]: Maybe an Optional return type instead of nullable?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeUpdatablePageSource.java [hunk 14]
```diff
@@ -574,7 +575,7 @@ private ReaderPageSource createParquetPageSource(TupleDomain<HiveColumnHandle> p
                 fileSize,
                 columns,
                 parquetPredicate,
-                true,
+                NAME,
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeUpdatablePageSource.java:578]: Can you add a TODO comment that explains that this is safe because we don't have writes to tables with column mappings enabled yet?

**Review by alexjo2144 — COMMENTED** [review]
Just skimming so far, but had one high level question.

Here we're passing the idea of column ids all the way through to the parquet reader. Would it be easier to leave the reader alone and use the existing index based approach? We could read the schema when setting up the reader and use the ids there to figure out the indexes in the Parquet file. 

Does that make sense?

**Comment by ebyhr** [comment]
> Would it be easier to leave the reader alone and use the existing index based approach?

Let me try this approach.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java [hunk 15]
```diff
@@ -151,7 +161,7 @@ public HiveColumnHandle toHiveColumnHandle()
     {
         return new HiveColumnHandle(
                 physicalName, // this name is used for accessing Parquet files, so it should be physical name
-                0, // hiveColumnIndex; we provide fake value because we always find columns by name
+                fieldId.orElse(0), // hiveColumnIndex; we provide fake value because we always find columns by name
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java:164]: I think with how you have this now you can undo this change and leave it as 0 all the time, right?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java [hunk 16]
```diff
@@ -134,6 +148,12 @@ public ConnectorPageSource createPageSource(
             }
         }
 
+        TrinoInputFile inputFile = fileSystemFactory.create(session).newInputFile(split.getPath(), split.getFileSize());
+        ParquetReaderOptions options = parquetReaderOptions.withMaxReadBlockSize(getParquetMaxReadBlockSize(session))
+                .withUseColumnIndex(isParquetUseColumnIndex(session));
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:153]: I guess you mixed up `isParquetUseColumnIndex` with `isUseParquetColumnNames`.

**Comment by ebyhr** [comment]
Addressed commnts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:153]: Yep, that's exactly what I did

**Comment by alexjo2144** [comment]
One last thing, but besides that looks good to me.

You don't have to store the column mapping mode in every split, since it is a table level property it can't change from file to file. Instead, the PageSourceProvider has access to the DeltaLakeTableHandle, you can get it from there

**Comment by ebyhr** [comment]
Thanks, updated.

**Review by alexjo2144 — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java [hunk 17]
```diff
@@ -194,7 +215,23 @@ public ConnectorPageSource createPageSource(
...
-    private static TupleDomain<HiveColumnHandle> getParquetTupleDomain(TupleDomain<DeltaLakeColumnHandle> effectivePredicate)
+    private Map<Integer, String> loadParquetIdAndNameMapping(TrinoInputFile inputFile, ParquetReaderOptions options)
+    {
+        try (ParquetDataSource dataSource = new TrinoParquetDataSource(inputFile, options, fileFormatDataSourceStats)) {
+            ParquetMetadata parquetMetadata = MetadataReader.readFooter(dataSource);
+            FileMetaData fileMetaData = parquetMetadata.getFileMetaData();
+            MessageType fileSchema = fileMetaData.getSchema();
+
+            return fileSchema.getFields().stream()
+                    .filter(field -> field.getId() != null)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:226]: Could you please document the purpose of this filter?

**Review by findepi — APPROVED** [review]
Skimming. 
Thank you @alexjo2144 @findinpath for your reviews!

**Comment by ebyhr** [comment]
CI hit #12818

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 18]
```diff
@@ -173,8 +177,88 @@ public void testColumnMappingModeNameWithNonLowerCaseColumnName()
...
+
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES  (3, 'new column')");
+            expectedRows = ImmutableList.of(
+                    row(1, null),
+                    row(2, null),
+                    row(3, "new column"));
+            assertThat(onDelta().executeQuery("SELECT * FROM default." + tableName))
+                    .containsOnly(expectedRows);
+            assertThat(onTrino().executeQuery("SELECT * FROM delta.default." + tableName))
+                    .containsOnly(expectedRows);
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:229]: While testing, I added the following snippet on your test case:

```
onDelta().executeQuery("ALTER TABLE default." + tableName + " DROP COLUMN another_varchar");
            assertThat(onTrino().executeQuery("DESCRIBE delta.default." + tableName))
                    .containsOnly(row("a_number", "integer", "", ""));
            expectedRows = ImmutableList.of(row(1), row(2), row(3));
            assertThat(onDelta().executeQuery("SELECT * FROM default." + tableName))
                    .containsOnly(expectedRows);
            assertThat(onTrino().executeQuery("SELECT * FROM delta.default." + tableName))
                    .containsOnly(expectedRows);
```


Since we are testing on how to the column mapping acts when adding a column we may as well verify how it acts when dropping a column.

I stumbled into an unrelated issue:

```
| 2022-09-07 20:21:44 INFO: FAILURE     /    io.trino.tests.product.deltalake.TestDeltaLakeColumnMappingMode.testColumnMappingModeIdAddColumn (Groups: profile_specific_tests, delta-lake-exclude-91, delta-lake-databricks, delta-lake-exclude-73) took 25.6 seconds
tests               | 2022-09-07 20:21:44 SEVERE: Failure cause:
tests               | io.trino.tempto.query.QueryExecutionException: java.sql.SQLException: [Databricks][DatabricksJDBCDriver](500051) ERROR processing query/statement. Error Code: 0, SQL state: org.apache.hive.service.cli.HiveSQLException: Error running query: java.lang.UnsupportedOperationException: Unrecognized column change class org.apache.spark.sql.connector.catalog.TableChange$DeleteColumn. You may be running an out of date Delta Lake version.
tests               | 	at org.apache.spark.sql.hive.thriftserver.HiveThriftServerErrors$.runningQueryError(HiveThriftServerErrors.scala:53)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:435)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.$anonfun$run$2(SparkExecuteStatementOperation.scala:257)
tests               | 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
tests               | 	at org.apache.spark.sql.hive.thriftserver.ThriftLocalProperties.withLocalProperties(ThriftLocalProperties.scala:123)
tests               | 	at org.apache.spark.sql.hive.thriftserver.ThriftLocalProperties.withLocalProperties$(ThriftLocalProperties.scala:48)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.withLocalProperties(SparkExecuteStatementOperation.scala:52)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:235)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2$$anon$3.run(SparkExecuteStatementOperation.scala:220)
tests               | 	at java.security.AccessController.doPrivileged(Native Method)
tests               | 	at javax.security.auth.Subject.doAs(Subject.java:422)
tests               | 	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
tests               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$2.run(SparkExecuteStatementOperation.scala:269)
tests               | 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
tests               | 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
tests               | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
tests               | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
tests               | 	at java.lang.Thread.run(Thread.java:748)
tests               | Caused by: java.lang.UnsupportedOperationException: Unrecognized column change class org.apache.spark.sql.connector.catalog.TableChange$DeleteColumn. You may be running an out of date Delta Lake version.
```

Will ask about it on Databricks slack

https://delta-users.slack.com/archives/CGK79PLV6/p1662561719425569

**Review by findinpath — APPROVED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 18 above)
Comment (alexjo2144) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:229]: You need to be running Databricks 11 for `DROP COLUMN` https://docs.databricks.com/release-notes/runtime/11.0.html#support-for-dropping-columns-in-delta-tables-public-preview

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java [hunk 19]
```diff
@@ -171,21 +197,21 @@ public ConnectorPageSource createPageSource(
...
                 fileFormatDataSourceStats,
-                parquetReaderOptions.withMaxReadBlockSize(getParquetMaxReadBlockSize(session))
-                        .withUseColumnIndex(isParquetUseColumnIndex(session)));
+                options);
 
         verify(pageSource.getReaderColumns().isEmpty(), "All columns expected to be base columns");
 
         return new DeltaLakePageSource(
                 deltaLakeColumns,
+                nullColumnNames.build(),
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:214]: I think you could get by without passing the null column names all the way through. The ParquetPageSource should already return nulls for columns whose name doesn't show up in the file. Maybe if the column is missing from the file you could give it a name like `missing_column_<uuid>`? 

I'm not sure if this is cleaner though.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 18 above)
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:229]: Indeed. Thanks @alexjo2144 . Testing with Databricks 11 the scenario evoked earlier proved to be successful.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 19 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:214]: Actually, it was my first approach and settled on the current change because I felt it's not straightforward and the `ParquetPageSource` behavior may change in the future.

@findepi Do you have any opinion?

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 19 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:214]: I am not being able to fully weight the consequences, but i agree that dummy names look weird.
Would would be the benefit of doing so?

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 19 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:214]: It gets rid of the changes in `DeltaLakePageSource`. That's about it

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSource.java [hunk 20]
```diff
@@ -112,8 +114,14 @@ else if (column.getName().equals(ROW_ID_COLUMN_NAME)) {
                 delegateIndex++;
             }
             else {
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSource.java:116]: Nit, if we're going to leave this in let's inline this if/else with the main one
```
else if (nullColumnNames.contains(...)) {
...
}
else {
...
}
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java (hunk 19 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java:214]: I don't really have a preference, happy to merge as is

**Comment by ebyhr** [comment]
Addressed all comments.

**Comment by ebyhr** [comment]
CI hit #13199 in `TestDeltaLakeDatabricksInsertCompatibility.testCompressionWithOptimizedWriter  `

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

**Review by alexjo2144 — APPROVED** [review]

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java [hunk 21]
```diff
@@ -57,26 +58,29 @@
 
     private final String name;
     private final Type type;
+    private final OptionalInt fieldId;
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeColumnHandle.java:61]: include in `getRetainedSizeInBytes` too

cc @krvikash
